opamp/observiq/observiq_client.go

```diff
@@ -229,6 +238,91 @@ func (c *Client) onRemoteConfigHandler(ctx context.Context, remoteConfig *protob
 	return nil
 }
 
+func (c *Client) onPackagesAvailableHandler(packagesAvailable *protobufs.PackagesAvailable) error {
+	c.logger.Debug("Packages available handler")
+
+	// Initialize PackageStatuses that will eventually be sent back to server
+	curPackageStatuses := &protobufs.PackageStatuses{
+		ServerProvidedAllPackagesHash: packagesAvailable.GetAllPackagesHash(),
+		Packages:                      map[string]*protobufs.PackageStatus{},
+	}
+
+	// Retrieve last known status (this should return with minimal info even on first time)
+	lastPackageStatuses, err := c.packagesStateProvider.LastReportedStatuses()
+
+	// If there is a problem retrieving the last saved PackageStatuses, we will log the error
+	// but continue on as the only thing missing will be the agent package hash.
+	if err != nil {
+		c.logger.Debug("Failed to retrieve last reported package statuses", zap.Error(err))
+	}
+
+	var lastPkgStatusMap map[string]*protobufs.PackageStatus
+	if lastPackageStatuses != nil {
+		lastPkgStatusMap = lastPackageStatuses.GetPackages()
+	}
+
+	curPackages := c.createPackageStatusMap(packagesAvailable.GetPackages(), lastPkgStatusMap)
+	curPackageStatuses.Packages = curPackages
+
+	if err = c.packagesStateProvider.SetLastReportedStatuses(curPackageStatuses); err != nil {
+		c.logger.Debug("Failed to save last reported package statuses", zap.Error(err))
+	}
+
+	if err = c.opampClient.SetPackageStatuses(curPackageStatuses); err != nil {
+		return fmt.Errorf("failed to set package statuses: %w", err)
+	}
+
+	return nil
+}
+
+func (c *Client) createPackageStatusMap(
+	pkgAvailMap map[string]*protobufs.PackageAvailable,
+	lastPkgStatusMap map[string]*protobufs.PackageStatus) map[string]*protobufs.PackageStatus {
+	pkgStatusMap := map[string]*protobufs.PackageStatus{}
+
+	// Loop through all of the available packages sent from the server
+	for name, availPkg := range pkgAvailMap {
+		switch name {
+		// If it's an expected package, return an installing status
+		case "observiq-otel-collector":
+			var agentHash []byte
+			if lastPkgStatusMap != nil && lastPkgStatusMap[name] != nil {
+				if lastPkgStatusMap[name].GetAgentHasVersion() != version.Version() {
+					c.logger.Debug(fmt.Sprintf(
+						"Version: %s and last reported package status version: %s differ",
+						version.Version(),
+						lastPkgStatusMap[name].GetAgentHasVersion()))
```

Comment:
Is this a place holder to where we would kick off update?

Reply:
@cpheps yeah to some degree. this built status implies that an update would actually be started (but i'm not implementing any of that with this story. just handling the PackageStatuses messages...which i know ultimately won't fully make sense without the update actually occurring)